orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py

```diff
@@ -0,0 +1,124 @@
+            which means ith gradient output, ith input, ith output of forward node respectively,
+            or any string for intermediate output of one of other nodes.
+        'outputs' is a list of strings. Each input string can be GI(i),
+            which means ith gradient input of forward node,
+            or any string for intermediate output for some of other nodes.
+        'attributes' (if present) is a dictionary. Each entry's key is the attribute name,
+            the value is also a dictionary: {'value': v, 'dtype': t, 'is_tensor': b}.
+                'v' can be string, number, or list of numbers, which is the value of the attribute.
+                't' is a string to describe the element type. It can be 'int', 'float', 'bool', etc,
+                    or 'IElemType(i)', 'OElemType(i)', which means the same type of ith input or
+                    output of forward node.
+                'b' is True or False, indicating if this attribute is tensor_proto ot not.
+                    'is_tensor' is optional, if not present, the default is False.
+"""
+
+import json
+from torch.onnx.symbolic_helper import _get_tensor_dim_size, _get_tensor_sizes
+
+
+def gradient_definition_to_json(gradient_def):
+    nodes = []
+    for node_def in gradient_def:
+        node_json = {}
+        # Op type
+        if isinstance(node_def[0], str):
+            node_json['op_type'] = node_def[0]
+        else:
+            node_json['op_type'] = node_def[0][0]
+            node_json['domain'] = node_def[0][1]
+        # Inputs and outputs
+        node_json['inputs'] = node_def[1]
+        node_json['outputs'] = node_def[2]
+        # Attributes
+        if len(node_def) >= 4:
+            node_json['attributes'] = node_def[3]
+        nodes.append(node_json)
+    return json.dumps(nodes)
+
+
+class ATenOpFactory:
+    _SYMBOLICS = {}
+    _GRADIENTS = {}
+
+    @classmethod
+    def register_symbolic(cls, name, domain=''):
+        def symbolic_wrapper(fn):
+            cls._SYMBOLICS[(name, domain)] = fn
+            return fn
+        return symbolic_wrapper
+
+    @classmethod
+    def register_gradient(cls, op_name, overload_name=''):
+        def gradient_wrapper(fn):
+            cls._GRADIENTS[(op_name, overload_name)] = gradient_definition_to_json(fn())
+            return fn
+        return gradient_wrapper
+
+
+@ATenOpFactory.register_symbolic('embedding')
+def embedding(g, weight, indices, padding_idx, scale_grad_by_freq, sparse):
+    output = g.op("com.microsoft::ATenOp", weight, indices, padding_idx, scale_grad_by_freq, sparse,
+                  name_s='aten::embedding')
+    indices_shape = _get_tensor_sizes(indices)
+    if indices_shape is not None and hasattr(weight.type(), 'with_sizes'):
+        output_type = weight.type().with_sizes(
+            indices_shape + [_get_tensor_dim_size(weight, 1)])
+        output.setType(output_type)
+    return output
+
+
+@ATenOpFactory.register_gradient('aten::embedding')
```

Comment:
this is really cool...
I think register_gradient method should have it's own class, as this can also be used for non-ATenOp

Reply:
However, this gradient builder doesn't have reference to the forward node. 
What if we wish to express some complex gradient building logic? e.g. check if broadcast handling is needed, check num_inputs in original node....